src/popup/services/cozyClient.service.ts

```
@@ -22,23 +28,48 @@ export class CozyClientService {
 
     getCozyURL(): string {
         const vaultUrl = this.environmentService.getWebVaultUrl();
-        return new URL(vaultUrl).origin; // Remove the /bitwarden part
+        if (!vaultUrl) {
+            return null
+        }
+        return new URL(vaultUrl).origin // Remove the /bitwarden part
+    }
+
+    async getClientInstance () {
+        try {
+            if (this.instance) {
+                const token = await this.apiService.getActiveBearerToken();
+                // If the instance's token differ from the active bearer, a refresh is needed.
+                if (token === this.instance.options.token) {
+                    return this.instance;
+                }
+            }
+            this.instance = await this.createClient();
+        } catch (err) {
+            /* tslint:disable-next-line */
+            console.error('Error while initializing cozy-client');
+            /* tslint:disable-next-line */
+            console.error(err);
+        }
+        return this.instance;
     }
 
     async createClient() {
-        if (this.instance) {
-            return this.instance;
+        try  {
+            const uri = this.getCozyURL();
+            const token = await this.apiService.getActiveBearerToken();
+            this.instance = new CozyClient({ uri: uri, token: token });
+        } catch (err) {
+            /* tslint:disable-next-line */
+            console.error('Error while initializing cozy-client');
+            /* tslint:disable-next-line */
+            console.error(err);
```

The exception handling for the creation of the client is done twice, here and in getClientInstance, is it necessary ? Why do we catch the exception here ?

---

I would answer: why not? We can call `createClient` directly, just like in runtime.background, so isn't better to have a specific exception handling here?

---

My point is that the exception is not really handled here, it is merely ignored. What should happen if the client creation fails ? The caller will have an undefined value returned right ? So in fact, by ignoring the exception, we are not really helping the caller since it will need to check if the client has been correctly created.

---

Hm ok I see your point; so, you think it would actually be better not catching errors here, then?

---

Yes. It's better to not swallow exception if we do not know how to deal with the error. A potential way to deal with errors would be to have an alert with "Something wrong happened and we do not know why, the extension is going to reload" with a global try/catch.

---

Here I think that the error is sufficiently grave for the extension to be reloaded. In case of a defective client, it's preferable to reload right ?

---

Ok for the error handling, I removed the trapping with 2dcb829